### DISCUSSION TITLE
Share routes, tracks, favorites, profiles with nearby devices

### DISCUSSION THREAD

**Discussion by Leeb64** [devemux86/cruiser#330] [Q&A] [answered]
A practical function would be sharing routes.
You are traveling with several people (motorcyclists, etc.). One of them plans a route and shares it with the others via WhatsApp, Signal, etc.
This can be done by saving the route in an external folder which you can then send.
It would be easier to send a route saved in Cruiser directly with a messenger.
Would something like that be possible?

**Comment by devemux86** [comment]
@Leeb64 thanks for the idea!

Indeed this would be useful.

Currently we can share routes (ready for offline navigation):
Export and share via cloud services, messengers, email, ...

Another idea could be the [Android Nearby](https://developers.google.com/nearby/):
_"Discover and establish direct communication channels with other devices without having to be connected to the Internet."_
_"Offline file transfers: Share photos, videos, or any other type of data quickly and without requiring a network connection."_

**Comment by SF4User** [comment]
Das ginge doch sooo einfach:
<img src="https://github.com/devemux86/cruiser/assets/132825580/859308cd-8f5c-4acf-8c24-0b4e0a64f47c" width="50%"/>

**Reply by devemux86** [reply]
@SF4User 

Do you mean QR codes?
They have a small size limit and are only suitable for small texts and links.
They cannot transfer complete routes **offline**.

Meinen Sie QR-Codes?
Sie haben eine kleine Größenbegrenzung und sind nur für kleine Texte und Links geeignet.
Sie können **offline** keine kompletten Routen übertragen.

- https://github.com/devemux86/cruiser/discussions/291

**Reply by SF4User** [reply]
Sorry! this is a QR-Screen Shot from Kurviger #2,
Your Partner bevore. Who is the Problem?
The Link to share have 18 Character, including the https Adress. The distribution of the CR-Code between a Group is a Action of Minutes.
Is not needet to share the complete Tour, only a schort Link to the Server. All Cruiser Customer have an Access to, after Load the Link or QR-Code.

**Reply by devemux86** [reply]
> Link to share

Kurviger links cannot be used in other apps and require internet to calculate routes.

Sharing routes offline is like sharing route files, no internet needed, just load and go.

**Reply by SF4User** [reply]
Wenn ich z.B. eine Kurviger Tour, zu Hause geplant und in den Cruiser importiert habe, ist das ja ab hier die Cruiser App und eine Cruiser Datei. Wenn nun unterwegs etwas geändert wird, dann sollte man doch analog zum Kurviger früher,  aus dem Cruiser einen Short Link oder einen QR-Code - nun dieser Cruiser Tour, aus dem Cruiser exportieren können. 
Was verstehe ich da nicht?
For example, if I have planned a Kurviger tour at home and imported it into the Cruiser, from here on it is the Cruiser app and a Cruiser file. If something is changed on the way, then you should be able to export a short link or a QR code from the Cruiser - just like with the Kurviger earlier - from the Cruiser - now this is Cruiser Tour.
What don't I understand?

**Reply by devemux86** [reply]
> export a short link or a QR code

Short links and QR codes contain full links to websites. It means they need internet to recalculate the routes.
Also these websites can only be used with their apps.

We are looking for a global sharing solution that can be used by everyone
and if it can work offline that will be a huge bonus.

**Reply by SF4User** [reply]
The shearing of tour links only works online. . . logical! The tour must first be calculated on the server. . . via from a central location (cloud). That would be the download link, you can then just download it easily, a cruiser (+) version and it costs (...) point.
Servers are expensive, the feature has to be paid for by the customer who benefits from it.

**Comment by 0709wiwiwi** [comment]
If a Locus user is also reading along here, he should take a moment to read the QR code. Even in the small format it should work. 
The first time you need to download and confirm with your browser and than select ALWAYS open in Locus. 
As one single poi wpt does not contain a Garmin compatible sym the Locus default wpt icon is shown.  
Inside that waypoint there is a linked local gpz photo image that can be set into a thumbnail icon.  
Only experienced Locus power users will have the simple know how to activate as thumbnail Icon I suppose. 
Anyway it could be offered automatically by simple means. Job to do by the app developer of course.

Besides the poi wpt, it also contains a ready-to-use navigation track so that you can start navigating in no time after import.

I did suggest and presented this in the Locus forum with a gif demo to see the nice final results but the Locus users are so enthusiastic and cooperative that 0 (zero) replies did follow.

<img src="https://github.com/devemux86/cruiser/assets/54167201/63395f27-8ea1-44c2-9097-f70f069939a7" width="30%"/>

**Comment by MotoUKRider** [comment]
Perhaps a Bluetooth file transfer (with the Cruiser GUI handling the handshake/security aspects) is the way to do this?
I imagine there would have to be some kind of PIN-accept on the receive-end to prevent "forced-push" transfers and other nasties.

**Reply by 0709wiwiwi** [reply]
A memory tip this topic has been mentioned once.
https://github.com/devemux86/cruiser/discussions/291

And I had already mentioned next promising development.

[QR code nearby sharing](https://en.shiftdelete.net/share-files-anywhere-with-nearby-shares-new-qr-code-support/#:~:text=Nearby%20Share%20gets%20QR%20code%20support&text=Tap%20the%20button%20to%20create,scanned%2C%20file%20transfer%20will%20begin.&text=The%20new%20QR%20code%20support%20is%20not%20yet%20available%20to%20users. )

I don't have an A14 but I haven't read anywhere that such has been realised. 
Google rather prefer to amuse users with even more file access hurdles.

**Comment by WolfgangZ42** [comment]
Sorry - where is the problem?
My workflow for sharing:
I load the route out of /My places/Routes, if not already in process anyway. Then /Routing/Export, having all options I prefer already shown, in my preferred folder - name of the route provided. Then with Datamanager+ picked up and shared via WhatsApp ... another few taps. 
And for groups travelling several days all members have access to a shared Dropbox folder each can copy routes in. Quite easy!

**Comment by devemux86** [comment]
- We can share routes, tracks, favorites, profiles with the new **Share** function

- Select **Quick Share** or **Bluetooth** to transfer offline content to nearby devices

Instructions of Quick Share (formerly Nearby Share) are available [here](https://support.google.com/android/answer/9286773) and [here](https://blog.google/products/android/how-to-use-quick-share-android/).
Instructions of Bluetooth are available [here](https://support.google.com/android/answer/9075925).

<img src="https://github.com/devemux86/cruiser/assets/3484020/10d0ecc7-09b3-4fba-9d75-7fb789df815c" width="33%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/6b07a948-79ca-4dbf-a958-2b17ce46b19b" width="33%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/3e59e2b2-7ab1-4fc5-b051-9ee5aa644a07" width="33%"/>

**Reply by umpfmpf** [reply]
The share menu looks very good. 
So I can share the file in all ways. Only the path from Android to Iphone is a bit bumpy. 
Otherwise it is also very nice to simply save the file.
Good idea!
Maybe you can completely replace the "Export" button with "Share"?

**Reply by devemux86** [reply]
Cruiser sends the file using official Android process, so these screens are from Android.

@umpfmpf have you tested "Quick Share" with 2 Android devices?
They mention that it can work offline with direct Wi-Fi connection.
Another alternative is using Bluetooth.

> Maybe you can completely replace the "Export" button with "Share"?

Indeed, although we need a way to save files locally?
Total Commander works, but I don't see Google Files (existing on all phones) in the list.

**Reply by umpfmpf** [reply]
I can test it tonight.

**Reply by bios4** [reply]
> Maybe you can completely replace the "Export" button with "Share"?

I ask to keep the Export button as is!
"Sharing" per se is something different than exporting a route/file in a specific format, so I would not change this...

**Reply by devemux86** [reply]
_This is just a first implementation for testing how the Quick Share or Bluetooth works between your devices._

The full implementation will display a similar screen (like export) to select format and options and then share the result.

**Reply by devemux86** [reply]
Also we can add similar "Share" action in favorites / routes / tracks and profiles.

**Reply by Ciutnik** [reply]
I did a quick test of Quick Share, away from WiFi. It took a while, the route.json file was saved in the Quick Share folder, the system suggested opening it, but instead of selecting the application, the system file browser opened. And here, after pointing to the file, QuickEdit+ (a cool text editor) opened without asking.
It's strange, because I don't remember associating .json files with any application... But this is, of course, behavior independent of Cruiser. He did his job correctly, sent the file correctly to the neighboring device 👍

Thanks for implementing this useful feature!

**Reply by devemux86** [reply]
@Ciutnik thanks for the test!

Probably the system does some automatic things or associations?

What happens if you (long) press the json in a file manager?
Doesn't it show the "open with" dialog?

**Reply by Ciutnik** [reply]
> Probably the system does some automatic things or associations?

Evidently yes! Selecting "Open with" from the menu also immediately opens QuickEdit+ without prompting. The situation is different on the second device (also Samsung). There, after clicking route.json, it proposes an application between QuickEdit+ and... the PowerAmp music player 🙂
That's why I do NOT use the Samsung system file manager on a daily basis. My favorite Solid Explorer allows any association of files with applications, so in addition to QuickEdit+ (which is most suitable for editing .json with syntax coloring) Cruiser is also available 🙂 And from then on it's been very good! 👍

**Reply by devemux86** [reply]
As seen in [above screenshots](https://github.com/devemux86/cruiser/discussions/330#discussioncomment-8782827) there is also Total Commander, that can save the shared file locally.

So this can cover the export too, but we need it to work without extra file manager.
Do you see the default file manager or Google Files in your phone in the share list?

**Reply by Ciutnik** [reply]
> Do you see the default file manager or Google Files in your phone in the share list?

Oh yes! I can see the entire set of applications on the device I'm sending from. The problems described above concern the receiving device. Unfortunately, after receiving the file and the "Open" option appears, we have no influence on what happens. This is the case at least in Samsung phones, which force you to use their file explorer here.

Is it possible for Cruiser to somehow register the .json file association in the system? Just like QuickEdit and Poweramp had to do...

**Reply by devemux86** [reply]
Cruiser is already registered for all file formats / extensions it uses.
But the final result really depends on Android and device changes.

> I can see the entire set of applications on the device I'm sending from.

This is what I mean: can you use the "Share" to save locally the route?
By selecting a file manager during the share process?

**Reply by umpfmpf** [reply]
It works fine with Dateimanager+ and X-plore, but not with the build in "Eigene Dateien" from Samsung (i cant see it in "Share")

**Reply by Ciutnik** [reply]
Fun fact. Among a very extensive list of applications - including, of course, file managers such as Solid Explorer and DS File from Synology - I do not see the default Samsung My Files (or something like that)...

However, there is a QR Code (!) as a Quick Share option, and it works - via Samsung servers.

**Comment by 0709wiwiwi** [comment]
The default android file explorer triggered by a Cruiser quick share transfer onto a moto G42 receiver offers only to open the json  file by the gpx viewer app altough the gpx viewer app does not import that format.

If you open the default android file explorer triggered by the Marc app and browse to the json file in the quick share folder than both Cruiser and gpx viewer are offered to open this file.

https://play.google.com/store/apps/details?id=com.marc.files

**Comment by devemux86** [comment]
File association is a complicated thing and new Android versions have more restrictions.

Quick Share and Bluetooth at least transfer the files to the other device's storage,
so apps can open them with their interface or any file manager after the transfer.

---

In CES 2024, it was announced that Google's Nearby Share would be merging with Samsung's Quick Share,
adopting the name of the latter: https://blog.google/products/android/ces-2024-android-updates/

Quick Share is available on Android 6 and later.

**Reply by bios4** [reply]
Regarding file associations you could try "Better open with" from the Play Store: https://play.google.com/store/apps/details?id=com.aboutmycode.betteropenwith

Regarding sharing: I hope that sharing is then finally possible also WITHOUT any remote server owned by a big tech company, where nobody knows what's happening with the files... 😉

**Reply by devemux86** [reply]
Quick Share by Google is mentioned to be done **offline** with direct Wi-Fi connection or even Bluetooth.
Whatever the devices find available if they are not connected to the internet.

**Comment by umpfmpf** [comment]
If I send a route from Samsung s24 to Doogee s88 I cannot assign the file to another program if I open it with the standard file manager. 
If I then open it with File Manager+ I can link it with Cruiser. 
However, when I open the file, I get an error message:

![20240314_195105](https://github.com/devemux86/cruiser/assets/105640513/30e26314-ef0d-445b-bf9d-e45085e95200)

**Reply by devemux86** [reply]
@umpfmpf do you use the latest beta?
This should be already fixed.

**Reply by umpfmpf** [reply]
Not on s88.... Now it works.
And i also can share with my pcloud Account... Without google or microsoft.

**Comment by devemux86** [comment]
Share routes, tracks, favorites, profiles with nearby devices is available for testing in:
- https://github.com/devemux86/cruiser/discussions/482

Instructions with images are mentioned [above](https://github.com/devemux86/cruiser/discussions/330#discussioncomment-8782827).

**Reply by bios4** [reply]
Just found out that the "Share" icon in the long-press route menu is different to the one in Main menu > Routing.

Could you please update it to the correct icon @devemux86?
Right now it is misleading showing the same three dots icon as the "More" tiles...

<img src="https://github.com/devemux86/cruiser/assets/145460147/d634c7f7-aac4-4d44-a1b2-436c358296a1" width="50%">

**Reply by devemux86** [reply]
I can change it.
Note that this menu may be named "Share" but opens a new screen to share location in various formats.

**Reply by bios4** [reply]
Nevertheless the icon should be the same from my POV.
I would appreciate if you could change it (but for sure this is a minor issue)...

**Reply by devemux86** [reply]
I can change it. :slightly_smiling_face:

**Comment by devemux86** [comment]
If we want to avoid the double Export / Share dialog, we can keep only the Share dialog
and **select a file manager** in the sharing process. This will then export the data as a file.

Interestingly, Total Commander is in the list of suggested apps, while Google Files is not!

<img src="https://github.com/devemux86/cruiser/assets/3484020/6b07a948-79ca-4dbf-a958-2b17ce46b19b" width="50%"/>

**Reply by 0709wiwiwi** [reply]
> If we want to avoid the double Export / Share dialog, we can keep only the Share dialog.

Simplification is fine but don't overdo it. Import and export fit perfectly together.

_With the unrestrained 12 panel menu proposal, you would have had to puzzle less to make everything fit nicely ;-)_
![image](https://github.com/devemux86/cruiser/assets/54167201/8453cc4c-1d2d-4346-b21d-ce9834c594f5)
This puzzling, on the other hand, is likely to recur permanently with the introduction of still new features.

**Reply by devemux86** [reply]
I mean technically it is possible to have only 1 export / share process.

It seems that's how Android expects us to do it, using a file manager.
Also, other map apps like OsmAnd seem to work this way now.

In any case we can keep what is most comfortable to use.

**Comment by devemux86** [comment, marked as answer]
Share routes, tracks, favorites, profiles with nearby devices is available in [Cruiser 4.0.0](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Reply by 1maxi2** [reply]
Mit dem "Teilen " Symbol würde ich erwarten das ich eine Route per Mail verschicken kann. Aber das funktioniert nicht. Was mache ich falsch ?

**Reply by ramondevic** [reply]
Primer et pregunta el tipus d'arxiu que vols compartir i després et demana el sistema de compartició: whatsapp, email, carpeta, quick share, etc

**Reply by devemux86** [reply]
@1maxi2 

Have you dragged the Share screen up to show all the suggested applications?

Haben Sie den Teilen-Bildschirm nach oben gezogen, um alle vorgeschlagenen Anwendungen anzuzeigen?

Gmail:
<img src="https://github.com/devemux86/cruiser/assets/3484020/c7104f0a-35c2-4309-adbc-0d9667ff4ed1" width="50%"/>

**Reply by 1maxi2** [reply]
Wenn ich lange auf die Route clicke kommt das " Teilen " Symbol "
Wenn ich auf das Symbol clicke kommt der Bildschirm wie am angehängten Foto ?
![Screenshot_20240504_211309_Cruiser](https://github.com/devemux86/cruiser/assets/124886589/61306a07-9b05-4174-b7f9-cf077b490f85)

**Reply by devemux86** [reply]
@1maxi2 

You are using an old version, please update to Cruiser 4.
Then use the menu "Routing | Export / Share".

Sie verwenden eine alte Version, bitte aktualisieren Sie auf Cruiser 4.
Verwenden Sie dann das Menü "Routing | Export / Teilen".

**Reply by 1maxi2** [reply]
Sorry für die Belästigung. Ich habe tatsächlich noch nicht die Version 4
Ich dachte das versenden per Mail würde schon seit längerer Zeit funktionieren
